tests/test-upgrade.py

```diff
@@ -0,0 +1,76 @@
+import os
+import time
+from validators import validate_dns, validate_dashboard, validate_storage, validate_ingress
+from subprocess import check_call
+from utils import microk8s_enable, wait_for_pod_state, microk8s_disable, wait_for_installation
+
+upgrade_from = os.environ.get('UPGRADE_MICROK8S_FROM', 'beta')
+# Have UPGRADE_MICROK8S_TO point to a file to upgrade to that file
+upgrade_to = os.environ.get('UPGRADE_MICROK8S_TO', 'edge')
+
+
+class TestUpgrade(object):
+    """
+    Validates a microk8s upgrade path
+    """
+
+    def test_upgrade(self):
+        """
+        Deploy, probe, upgrade, validate nothing broke.
+
+        """
+        print("Testing upgrade from {} to {}".format(upgrade_from, upgrade_to))
+
+        cmd = "sudo snap install microk8s --classic --{}".format(upgrade_from).split()
+        check_call(cmd)
+        wait_for_installation()
+
+        # Run through the validators and
+        # select those that were valid for the original snap
+        test_matrix = {}
+        try:
+            microk8s_enable("dns")
+            wait_for_pod_state("", "kube-system", "running", label="k8s-app=kube-dns")
+            validate_dns()
+            test_matrix['dns'] = validate_dns
+        except:
+            print('Will not test dns')
```

Comment:
If `microk8s_enable('dns')` fails, that seems like it should be considered a failure and not a skip.   If the intent is to allow for a particular add-on being removed in the future, maybe the condition for doing the check should instead use the new `--help` option and check for the addon in the list?

Reply:
As mentioned in the previous comment we skip the test and we do not add it in the `test_matrix` because we are not interested in testing it after the upgrade.

Something else, the validation tests do not have to map to the addons we have. We may have a validation that deploys microbot and checks it is still up after an upgrade. I am mostly concerned with releasing to beta. Edge backwards compatibility is also nice, but it is beta where our users live at the moment.

Also all validations are already checked against the current snap here: https://github.com/juju-solutions/microk8s/blob/master/.travis.yml#L17